### PR TITLE
tls-passthrough docs and quickstart additions

### LINKED ISSUE #554: Document TLS Passthrough
Create a doc that shows a user how to configure TLS passthrough.

### PR DESCRIPTION
**Pull request by chauhanshubham**
Signed-off-by: Shubham Chauhan <shubham@tetrate.io>
Resolves #554

### CODE REVIEW

**Comment by chauhanshubham** [comment]
Need some suggestion around the quickstart.yaml additions. Should we add the TLSRoute specific configs in the same quickstart.yaml, or create a separate config to be applied for TLSRoute.

File: examples/kubernetes/quickstart.yaml [hunk 1]
```diff
@@ -15,6 +15,12 @@ spec:
     - name: http
       protocol: HTTP
       port: 8080
+    - name: tls
+      protocol: TLS
+      hostname: passthrough.example.com
+      tls:
+        mode: Passthrough
+      port: 8443
```
Comment (danehans) [inline examples/kubernetes/quickstart.yaml:23]: Due to https://github.com/envoyproxy/gateway/issues/520 and https://github.com/envoyproxy/gateway/pull/542 please use a different port number.

File: docs/user/TLS_PASSTHROUGH.md [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+Store the cert/keys in A Secret:
+```shell
+kubectl create secret tls nginx-server-certs --key=passthrough.example.com.key --cert=passthrough.example.com.crt
+```
+
+### Application config
+
+Create a file for the nginx server:
+```shell
+cat <<\EOF > ./nginx-passthrough.conf
```
Comment (danehans) [inline docs/user/TLS_PASSTHROUGH.md:48]: The plan is to get rid of httpbin and consolidate all examples to use the same kube conformance echo server. The echo server [supports TLS](https://github.com/kubernetes-sigs/ingress-controller-conformance/blob/master/images/echoserver/echoserver.go#L108-L123). Refer to https://github.com/envoyproxy/gateway/pull/558/ for additional details regarding the echo server.

**Review by danehans — CHANGES_REQUESTED** [review]

**Comment by danehans** [comment]
> Need some suggestion around the quickstart.yaml additions. Should we add the TLSRoute specific configs in the same quickstart.yaml, or create a separate config to be applied for TLSRoute.

As you have done, reference the quickstart to establish a baseline. Instead of adding the additional resources to the quickstart manifest, create a separate manifest for the Service, Deployment, and TLSRoute resources.

File: examples/kubernetes/quickstart.yaml (hunk 1 above)
Comment (chauhanshubham) [reply examples/kubernetes/quickstart.yaml]: using 6443 now

File: docs/user/tls-passthrough.md [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    },
+}]'
+```
+
+## Testing
+### Clusters without External LoadBalancer Support
+Port forward to the Envoy service:
+
+```shell
+kubectl -n envoy-gateway-system port-forward service/envoy-default-eg 8888:6443 &
```
Comment (danehans) [inline docs/user/tls-passthrough.md:76]: We are now hashing the names of the managed k8s resources. You will need:

```shell
export ENVOY_SERVICE=$(kubectl get svc -n envoy-gateway-system --selector=gateway.envoyproxy.io/owning-gateway-namespace=default,gateway.envoyproxy.io/owning-gateway-name=eg -o jsonpath='{.items[0].metadata.name}')
```

Port forward to the Envoy service:

```shell
kubectl -n envoy-gateway-system port-forward service/${ENVOY_SERVICE} 6043:6443 &
```

File: docs/user/tls-passthrough.md [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+```
+
+Curl the example app through Envoy proxy:
+
+```shell
+curl -v --resolve "passthrough.example.com:8888:127.0.0.1" https://passthrough.example.com:8888 \
+--cacert passthrough.example.com.crt
```
Comment (danehans) [inline docs/user/tls-passthrough.md:83]: `8888` is already being used by the quickstart HTTP service, so I replaced it with `6043` in my comment above.

File: docs/user/tls-passthrough.md [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+You can also test the same functionality by sending traffic to the External IP of the Gateway:
+
+```shell
+export GATEWAY_HOST=$(kubectl get gateway/eg -o jsonpath='{.status.addresses[0].value}')
+```
+
+Curl the example app through the Gateway, e.g. Envoy proxy:
+
+```shell
+curl -v -HHost:passthrough.example.com --resolve "passthrough.example.com:8888:${GATEWAY_HOST}" \
```
Comment (danehans) [inline docs/user/tls-passthrough.md:96]: s/8888/6443/

```
$ curl -v -HHost:passthrough.example.com --resolve "passthrough.example.com:6443:${GATEWAY_HOST}" \
--cacert example.com.crt https://passthrough.example.com:6443/get
* Added passthrough.example.com:6443:172.18.255.200 to DNS cache
* Hostname passthrough.example.com was found in DNS cache
*   Trying 172.18.255.200:6443...
* Connected to passthrough.example.com (172.18.255.200) port 6443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*  CAfile: example.com.crt
*  CApath: none
* (304) (OUT), TLS handshake, Client hello (1):
* (304) (IN), TLS handshake, Server hello (2):
* (304) (IN), TLS handshake, Unknown (8):
* (304) (IN), TLS handshake, Certificate (11):
* (304) (IN), TLS handshake, CERT verify (15):
* (304) (IN), TLS handshake, Finished (20):
* (304) (OUT), TLS handshake, Finished (20):
* SSL connection using TLSv1.3 / AEAD-AES128-GCM-SHA256
* ALPN, server accepted to use h2
* Server certificate:
*  subject: CN=passthrough.example.com; O=some organization
*  start date: Oct 18 17:55:26 2022 GMT
*  expire date: Oct 18 17:55:26 2023 GMT
*  common name: passthrough.example.com (matched)
*  issuer: O=example Inc.; CN=example.com
*  SSL certificate verify ok.
* Using HTTP2, server supports multiplexing
* Connection state changed (HTTP/2 confirmed)
* Copying HTTP/2 data in stream buffer to connection buffer after upgrade: len=0
* Using Stream ID: 1 (easy handle 0x7f9617012800)
> GET /get HTTP/2
> Host:passthrough.example.com
> user-agent: curl/7.79.1
> accept: */*
>
* Connection state changed (MAX_CONCURRENT_STREAMS == 250)!
< HTTP/2 200
< content-type: application/json
< x-content-type-options: nosniff
< content-length: 395
< date: Tue, 18 Oct 2022 18:08:35 GMT
<
{
 "path": "/get",
 "host": "passthrough.example.com",
 "method": "GET",
 "proto": "HTTP/2.0",
 "headers": {
  "Accept": [
   "*/*"
  ],
  "User-Agent": [
   "curl/7.79.1"
  ]
 },
 "namespace": "",
 "ingress": "",
 "service": "",
 "pod": "",
 "tls": {
  "version": "TLSv1.3",
  "serverName": "passthrough.example.com",
  "negotiatedProtocol": "h2",
  "cipherSuite": "TLS_AES_128_GCM_SHA256"
 }
```

File: examples/kubernetes/tls-passthrough.yaml [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    metadata:
+      labels:
+        run: passthrough-echoserver
+    spec:
+      containers:
+        - name: passthrough-echoserver
+          image: gcr.io/k8s-staging-ingressconformance/echoserver:v20220815-e21d1a4
+          ports:
+            - containerPort: 8443
+          env:
```
Comment (danehans) [inline examples/kubernetes/tls-passthrough.yaml:34]: Can you add the pod and namespace env vars too?

xref: https://github.com/envoyproxy/gateway/blob/main/examples/kubernetes/quickstart.yaml#L62-L70

**Review by danehans — CHANGES_REQUESTED** [review]
A few minor changes needed and then this should be good to go.

File: docs/user/tls-passthrough.md [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+For the application, we'll deploy a sample echoserver app, with the certificates loaded in the application Pod.
+
+__Note:__ These certificates will not be used by the Gateway, but will remain in the application scope.
+
+For macOS users, verify curl is compiled with the LibreSSL library:
+
+```shell
+curl --version | grep LibreSSL
+curl 7.54.0 (x86_64-apple-darwin17.0) libcurl/7.54.0 LibreSSL/2.0.20 zlib/1.2.11 nghttp2/1.24.0
+```
```
Comment (danehans) [inline docs/user/tls-passthrough.md:27]: Let's drop this step.

**Comment by chauhanshubham** [comment]
addressed the comments @danehans

File: docs/user/tls-passthrough.md [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+Follow the steps from the [Quickstart Guide](QUICKSTART.md) to uninstall Envoy Gateway and the example manifest.
+
+Delete the Secret:
+
+```shell
+kubectl delete secret/server-certs
+```
+
+## Next Steps
+Checkout the [Developer Guide](../../DEVELOPER.md) to get involved in the project.
```
Comment (danehans) [inline docs/user/tls-passthrough.md:111]: This doc moved to https://github.com/envoyproxy/gateway/blob/main/docs/dev/README.md. I'll approve and send a PR to fix it.

**Review by danehans — APPROVED** [review]

**Pull request merged**